feat: BlockData PREVRANDAO and BLOBBASEFEE

letypequividelespoubelles opened the pull request:
This PR:
- finishes the Paris HF with the implementation of PREVRANDAO instead of DIFICULTY
- Starts the Cancun HF with:
- implementation of BLOBBASEFEE opcode
- update the flag in CancunOpcode.yml for batch family opcode.

letypequividelespoubelles committed:
feat: Paris HF

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
feat: trace column fork dependant block data

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2037-finish-paris-hf

letypequividelespoubelles committed:
simplify tracing fork dependant columns in Oob/Hub

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
update opcode.yml for prevrandao

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
add Paris in corset validator

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix: opcodes list in blockdtata is fork dependant

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
deal with prevrandao for post paris

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
make it compile

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/BlockdataOperation.java:
```diff
@@ -113,27 +113,13 @@ public BlockdataOperation(
+      case COINBASE -> handleCoinbase();
+      case TIMESTAMP -> handleTimestamp();
+      case NUMBER -> handleNumber();
+      case DIFFICULTY, PREVRANDAO -> handleDifficultyOrPrevRandao(); // depending on the fork
```
I think it would be cleaner to have two separate methods
```java
case DIFFICULTY -> handleDifficulty();
case PREVRANDAO -> handlePrevRandao();
```
where both methods have a default implementation of throwing a runtime exception.

letypequividelespoubelles replied:
makes sense ... Idk why I wanted to have it in the same method :)

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/BlockdataOperation.java:
```diff
@@ -204,6 +190,8 @@ private void handleBaseFee() {
     wcpCallToGEQ(0, data, EWord.ZERO);
   }
 
+  protected abstract void traceIsDifficultyOrIsPrevRandao(Trace.Blockdata trace, OpCode opCode);
```
Similarly here, wouldn't we want two methods that by default do nothing to the trace ?

OlivierBBB commented on arithmetization/src/test/java/net/consensys/linea/zktracer/statemanager/StateManagerUtils.java:
```diff
@@ -33,7 +32,7 @@ public static int getTxCount(Hub hub) {
 
   public static int getBlockOperationsLength(Hub hub) {
-    return Blockdata.getOpCodes().length;
+    return 7; // = Blockdata.getOpCodes().length;
```
Merci ^^

OlivierBBB replied:
Though I wonder whether that is the most future proof solution. We will be adding a new `BLOCK_DATA` opcode for ~Prague I believe~ Cancun called `BLOBBASEFEE` which would automatically be accounted for if we kept the old pattern.

letypequividelespoubelles replied:
yes, because of BLOBBASEFEE we'll have an issue. The things is all those tests are static method (didn't check why) so it'll be annoying in the future. We'll see when implementing BLOBBASEFEE

OlivierBBB commented on gradle/corset.gradle:
```diff
@@ -40,6 +42,7 @@ tasks.register('buildZkevmBins', Exec) {
     inputs.files(fileTree(workingDir).include('**/*.lisp'))
     // Specify output file so gradle knows when out-of-date.
     outputs.files(fileTree(workingDir).include(ZKEVM_LONDON_BIN))
+    outputs.files(fileTree(workingDir).include(ZKEVM_PARIS_BIN))
```
A general remark is that for the constraints you have chosen 3 letter abbreviations for forks (e.g. LON, PAR) while here we use full names. Full names everywhere would seem preferable on all accounts.

letypequividelespoubelles replied:
I'm in favor of having the full name everywhere too.
@amkCha @DavePearce  thought ?

OlivierBBB replied:
I believe I discussed this with @amkCha some time ago in person and we agreed on this.

amkCha replied:
Hm, not sure where we use `LON`and `PAR` in the constraints anymore after all the iterations. The folders are using the full name, and the binary files as well. You talking about the variables in the Makefile ? If so, yes, we can change the names no problem https://github.com/Consensys/linea-constraints/blob/master/Makefile#L18

OlivierBBB reviewed (changes requested):
LGTM overall with some possible requested changes.

amkCha commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/fragment/imc/oob/opcodes/create/ShanghaiCreateOobCall.java:
```diff
@@ -43,25 +43,11 @@ protected void codeSizeSnapshot(final MessageFrame frame) {
-    } catch (Exception e) {
-      throw new IllegalArgumentException("Trace argument is not of type TraceShanghai.Oob", e);
-    }
+    ((TraceShanghai.Oob) trace).data10(Bytes.ofUnsignedLong(codeSize));
```
Much better, and you got rid of the `var` :+1:

letypequividelespoubelles committed:
clean: make separate method for different opcode (OB's comment)

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Merge branch 'arith-dev' into 2037-finish-paris-hf

letypequividelespoubelles committed:
almost done with blobbasefee in blockdata

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
ras

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

DavePearce committed:
support shared Paris/Shanghai interface

This adds support for building a shared interface between the Paris and
Shanghai forks.  This extends the root interface, and provides access to
features which are only present in those forks (i.e. which are not
present in others).

letypequividelespoubelles committed:
blobbasefee from frame

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
feat: update opcode.yml flag for batch opcode + rm one column in cancun block data

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
Trace.java will have all fork columns, no need to cast the trace

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix fork dependant block opcodes size

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

DavePearce committed:
Convert to "union of all traces" approach

This converts the trace generation process using go-corset to adopt the
"union of all traces approach" which produces just one Trace interface
which includes *everything*.

At this stage, there remains a problem with BlockData.nROWS_DEPTH which
was being statically imported.  However, this constant differs between
forks and, hence, is not visible in the Trace interface.

letypequividelespoubelles committed:
fix: ctMax per opcode

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
some cleaning

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix static tests by skipping non london opcodes

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
cleaning

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
fix tests

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
update gas projector with new opcodes

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/module/ParisBlockData.java:
There is no `ShanghaiBlockData` module as this module is the same as `Paris` I take it ?

letypequividelespoubelles replied:
exactly. The ShanghaiHub will create a ParisBlockData

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/BlockdataOperation.java:
```diff
@@ -166,7 +157,11 @@ private void handleNumber() {
+
+  protected abstract void handlePrevrandao();
+
+  protected abstract void handleBlobbasefee();
```
👍

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/BlockdataOperation.java:
```diff
@@ -241,11 +240,18 @@ public void trace(Trace.Blockdata trace) {
+
+  protected abstract void traceIsPrevrandao(Trace.Blockdata trace, OpCode opCode);
+
+  protected abstract void traceIsBlobbasefee(Trace.Blockdata trace, OpCode opCode);
```
Given what the execution specs use (https://github.com/ethereum/execution-specs/blob/81ec8eee0c8657bcc45fc0f979a8bbdaf868ddb1/src/ethereum/cancun/vm/instructions/__init__.py#L277), that is
```python
blob_base_fee
prev_randao
base_fee
```
It would make more sense to have method names
```hava
traceIsPrevRandao
traceIsBlobBaseFee
```
and actually rename the `BTC` column to `BASE_FEE`, too.

letypequividelespoubelles replied:
done

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/CancunBlockDataOperation.java:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  @Override
+  protected void handleBlobbasefee() {
```
Same camel case comment

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/LondonBlockDataOperation.java:
```diff
@@ -38,10 +41,40 @@ public LondonBlockDataOperation(
+  }
+
+  @Override
+  protected void traceIsPrevrandao(Trace.Blockdata trace, OpCode opCode) {
```
Same camel case comment

OlivierBBB commented on arithmetization/src/test/java/net/consensys/linea/zktracer/exceptions/OutOfGasExceptionTest.java:
```diff
@@ -81,6 +83,7 @@ void outOfGasExceptionWithEmptyAccountsAndNoMemoryExpansionCostTest(
 
   static Stream<Arguments> outOfGasExceptionWithEmptyAccountsAndNoMemoryExpansionCostTestSource() {
     List<Arguments> arguments = new ArrayList<>();
+    loadOpcodes(LONDON);
```
@letypequividelespoubelles What is the purpose of this ?

letypequividelespoubelles replied:
this is how we populate the (static) opcodeDataList from the ForkOpcodes.yml

OlivierBBB commented on arithmetization/src/main/resources/cancunOpcodes.yml:
Is this for now just a copy of `shanghaiOpcodes` with `BLOBBASEFEE` & `BLOBHASH` ? I.e. I can't e.g. find `MCOPY`.

letypequividelespoubelles replied:
for the moment, it has only the PREVRANDAO and the update for the BTC family opcode coming from the PR of BLOBBASEFEE. So yes, no transient, no MCOPY, etc, that will need to be added in relevant PRs

OlivierBBB reviewed (approved):
LGTM with minor naming convention comments

letypequividelespoubelles committed:
renaming

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
empty commit

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>

letypequividelespoubelles committed:
empty commit

Signed-off-by: F Bojarski <ceciestunepoubelle@protonmail.ch>
